crates/ruma-server-util/src/authorization.rs

```diff
@@ -0,0 +1,198 @@
+    pub destination: Option<Box<ServerName>>,
+    ///the ID, including the algorithm name, of the sending server’s key used to sign the request
+    pub key: Box<ServerSigningKeyId>,
+    /// the signature of the JSON
+    pub sig: String,
+}
+
+/// Builder for [`XMatrix`]
+pub struct XMatrixBuilder {
+    origin: Option<Box<ServerName>>,
+    destination: Option<Box<ServerName>>,
+    key: Option<Box<ServerSigningKeyId>>,
+    sig: Option<String>,
+}
+
+impl XMatrixBuilder {
+    /// Create an empty builder instance
+    pub fn new() -> Self {
+        Self { origin: None, destination: None, key: None, sig: None }
+    }
+
+    /// Set the origin attribute
+    pub fn origin(mut self, origin: Box<ServerName>) -> Self {
+        if self.origin.is_some() {
+            debug!("Field origin duplicated in X-Matrix Authorization header");
+        }
+        self.origin = Some(origin);
+        self
+    }
+
+    /// Set the optional destination attribute
+    pub fn destination(mut self, destination: Box<ServerName>) -> Self {
+        if self.destination.is_some() {
+            debug!("Field destination duplicated in X-Matrix Authorization header");
+        }
+        self.destination = Some(destination);
+        self
+    }
+
+    /// Set the key attribute
+    pub fn key(mut self, key: Box<ServerSigningKeyId>) -> Self {
+        if self.key.is_some() {
+            debug!("Field key duplicated in X-Matrix Authorization header");
+        }
+        self.key = Some(key);
+        self
+    }
+
+    /// Set the sig attribute
+    pub fn sig(mut self, sig: String) -> Self {
+        if self.sig.is_some() {
+            debug!("Field sig duplicated in X-Matrix Authorization header");
+        }
+        self.sig = Some(sig);
+        self
+    }
+
+    /// Build an [`XMatrix`] instance
+    pub fn build(self) -> Option<XMatrix> {
+        Some(XMatrix {
+            origin: self.origin?,
+            destination: self.destination,
+            key: self.key?,
+            sig: self.sig?,
+        })
+    }
+}
+
+impl Default for XMatrixBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Credentials for XMatrix {
+    const SCHEME: &'static str = "X-Matrix";
+
+    fn decode(value: &HeaderValue) -> Option<Self> {
+        debug_assert!(
+            value.as_bytes().starts_with(b"X-Matrix "),
+            "HeaderValue to decode should start with \"X-Matrix ..\", received = {:?}",
+            value,
+        );
+
+        let parameters =
+            std::str::from_utf8(&value.as_bytes()["X-Matrix ".len()..]).ok()?.trim_start();
+
+        let mut builder: XMatrixBuilder = Default::default();
+
+        for entry in parameters.split_terminator(',') {
+            let (name, value) = entry.split_once('=')?;
+
+            // It's not at all clear why some fields are quoted and others not in the spec,
+            // let's simply accept either form for every field.
```

Comment:
Please get some clarity on this and encode it into the logic. IIRC there was some debate surrounding this, either in an issue or in the spec room. If there is an issue, please try to find a conservative approach to parse this, and reference the issue itself.

If there is a defined way of parsing these quotes, please remove this comment, or clarify via which way (IIRC an RFC was mentioned) this is "alright" to parse.

Reply:
Yeah, I haven't looked at this closer, the spec is actually clear on this by now. TL;DR: If it's a valid token as defined in RFC 7230, quotes are allowed but not required, if it's not a valid token, but would be if it didn't contain colons, quotes aren't required either, but only for compatibility reasons. If it's not a valid token even without additionally allowing colons, quotes are a hard requirement.

How should this be validated? A regex? Would be fairly easy to do, but I feel like that might slow down the validation here a bit and am not sure whether it's worth it

Reply:
I mainly called out the RFC mention so you could find a valid way to parse this in one maintained (or std) crate or another, maybe in the `headers` crate?

Reply:
So, it's a bit complicated, and I'm not 100% sure how exactly this is supposed to work. Moving this upstream into the `headers` crate is probably not going to work because of this matrix spec part:

> For compatibility with older servers, the recipient should allow colons to be included in values without requiring the value to be enclosed in quotes.

So, if we stick to having our own parsing of it here, I think something along the lines of what I've built with `yap` here is the most correct solution we're going to get. It's certainly quite a bit more correct than the naive implementation I had cargo-culted from Conduit.

Reply:
After thinking about this some more: Maybe it'd be okay to skip this **should** requirement due to the relevant synapse instances being fairly out of date and vulnerable anyway. Without this should requirement an implementation for parsing credentials based on auth_params generically could be contributed into the `headers` crate upstream. Although, from my experience with https://github.com/hyperium/headers/pull/93, I'm not sure I'd get my hopes up about such a contribution being merged any time soon.